Source/processcontainers/implementations/CRunImplementation/CRunImplementation.cpp

```diff
@@ -0,0 +1,460 @@
+            if (ret < 0) {
+                TRACE_L1("Failed to acquire container \"%s\" state string", _name.c_str());
+                result = false;
+            } else {
+                result = (ret == 1);
+            }
+
+            libcrun_free_container_status(&status);
+        }
+
+        return result;
+    }
+
+    bool CRunContainer::Start(const string& command, IStringIterator& parameters) 
+    {
+        libcrun_error_t error = nullptr;
+        int ret = 0;
+        bool result = true;
+
+        // Make sure no leftover container instances are present
+        if (ClearLeftovers() != Core::ERROR_NONE) {
+            result = false;
+        }
+        else {
+            _container = libcrun_container_load_from_file (_configFile.c_str(), &error);
+
+            // Add bundle prefix to rootfs location if relative path is provided
+            // TODO: Possibly change mount relative path to acomodate for bundle or check if doing chdir() is safe alternative
+            string rootfsPath = _bundle + "/";
+            if (_container->container_def->root->path != nullptr && _container->container_def->root->path[0] != '/') {
+                rootfsPath += _container->container_def->root->path;
+                _container->container_def->root->path = 
+                    reinterpret_cast<char*>(realloc (_container->container_def->root->path, (rootfsPath.length() + 1) * sizeof(char)));
+                
+                strcpy(_container->container_def->root->path, rootfsPath.c_str());
+            }
+            
+            if (_container == NULL) {
+
+                TRACE_L1("Failed to load a configuration file in %s. Error %d: %s", _configFile.c_str(), error->status, error->msg);
+                result = false;
+            } else {
+
+                OverwriteContainerArgs(_container, command, parameters);
+
+                ret = libcrun_container_run (&_context, _container, LIBCRUN_RUN_OPTIONS_PREFORK, &error);
+                if (ret != 0) {
+                    TRACE_L1("Failed to run a container \"%s\". Error %d: %s", _name.c_str(), error->status, error->msg);
+                    result = false;
+                } else {
+                    _created = true;
+                }
+            }
+        }
+
+        return result;
+    }
+
+    bool CRunContainer::Stop(const uint32_t timeout /*ms*/)
+    {
+        bool result = true;
+        libcrun_error_t error = nullptr;
+
+        if (libcrun_container_delete (&_context, NULL, _name.c_str(), true, &error) != 0) {
+            TRACE_L1("Failed to stop a container \"%s\". Error: %s", _name.c_str(), error->msg);
+            result = false;
+        }
+        _created = false;
+
+        return result;
+    }
+
+    void CRunContainer::AddRef() const 
+    {
+        _refCount++;
+    };
+
+    uint32_t CRunContainer::Release()
+    {
+        if (--_refCount == 0) {
+            if (_created == true) {
```

Comment:
I think this container will never reach a reference count of 0 as the CRunAdministrator will always hold the last refernce..... It will only be cleared if the administrator is destructed. Is that as designed ? Should the container only stop if the ContainerAdministrator is destructed?